privacy/bolton/loss.py

```diff
@@ -0,0 +1,280 @@
+               name: str = None,
+               dtype=tf.float32,
+               **kwargs):
+    """
+    Args:
+      reg_lambda: Weight regularization constant
+      c: Additional constant for strongly convex convergence. Acts
+          as a global weight.
+      radius_constant: constant defining the length of the radius
+      reduction: reduction type to use. See super class
+      name: Name of the loss instance
+      dtype: tf datatype to use for tensor conversions.
+    """
+    super(StrongConvexLoss, self).__init__(reduction=reduction,
+                                           name=name,
+                                           **kwargs)
+    self._sample_weight = tf.Variable(initial_value=c,
+                                      trainable=False,
+                                      dtype=tf.float32)
+    self._reg_lambda = reg_lambda
+    self.radius_constant = tf.Variable(initial_value=radius_constant,
+                                       trainable=False,
+                                       dtype=tf.float32)
+    self.dtype = dtype
+
+  def radius(self):
+    """Radius of R-Ball (value to normalize weights to after each batch)
+
+    Returns: radius
+
+    """
+    raise NotImplementedError("Radius not implemented for StrongConvex Loss"
+                              "function: %s" % str(self.__class__.__name__))
+
+  def gamma(self):
+    """ Gamma strongly convex
+
+    Returns: gamma
+
+    """
+    raise NotImplementedError("Gamma not implemented for StrongConvex Loss"
+                              "function: %s" % str(self.__class__.__name__))
+
+  def beta(self, class_weight):
+    """Beta smoothess
+
+    Args:
+      class_weight: the class weights used.
+
+    Returns: Beta
+
+    """
+    raise NotImplementedError("Beta not implemented for StrongConvex Loss"
+                              "function: %s" % str(self.__class__.__name__))
+
+  def lipchitz_constant(self, class_weight):
+    """ L lipchitz continuous
+
+    Args:
+      class_weight: class weights used
+
+    Returns: L
+
+    """
+    raise NotImplementedError("lipchitz constant not implemented for "
+                              "StrongConvex Loss"
+                              "function: %s" % str(self.__class__.__name__))
+
+  def reg_lambda(self, convert_to_tensor: bool = False):
+    """ returns the lambda weight regularization constant, as a tensor if
+    desired
+
+    Args:
+      convert_to_tensor: True to convert to tensor, False to leave as
+                            python numeric.
+
+    Returns: reg_lambda
+
+    """
+    if convert_to_tensor:
+      return _ops.convert_to_tensor_v2(self._reg_lambda, dtype=self.dtype)
+    return self._reg_lambda
+
+  def max_class_weight(self, class_weight):
+    class_weight = _ops.convert_to_tensor_v2(class_weight, dtype=self.dtype)
+    return tf.math.reduce_max(class_weight)
+
+
+class Huber(StrongConvexLoss, losses.Huber):
```

Comment:
The loss function we implemented is different than any of those implemented in tensorflow. The paper has the formula.